server/src/api/util/level.js

```diff
@@ -0,0 +1,72 @@
+const _ = require('lodash');
+const { MAX_LEVEL, MAX_VIRTUAL_LEVEL } = require('../constants/levels');
+
+function getLevel(experience, virtual = false) {
+  // Unranked
+  if (experience === -1) {
+    return 0;
+  }
+
+  const maxlevel = virtual ? MAX_VIRTUAL_LEVEL : MAX_LEVEL;
+
+  let accumulated = 0;
+
+  for (let level = 1; level < maxlevel; level++) {
+    const required = getXpDifferenceTo(level + 1);
+    if (experience >= accumulated && experience < accumulated + required) {
+      return level;
+    }
+    accumulated += required;
+  }
+
+  return maxlevel;
+}
+
+function getXpDifferenceTo(level) {
+  if (level < 2) {
+    return 0;
+  }
+
+  return Math.floor(level - 1 + 300 * 2 ** ((level - 1) / 7)) / 4;
+}
+
+function getCombatLevel(playerExperiences) {
+  const combatExperiences = _.pick(playerExperiences, [
+    'attackExperience',
+    'strengthExperience',
+    'defenceExperience',
+    'hitpointsExperience',
+    'rangedExperience',
+    'prayerExperience',
+    'magicExperience'
+  ]);
+
+  const levels = _.transform(combatExperiences, (r, v, k) => {
+    // eslint-disable-next-line no-param-reassign
+    r[k.replace('Experience', '')] = getLevel(v);
+  });
+
+  // Ensure all skills are set
+  if (Object.keys(levels).length !== 7) {
```

Comment:
I don't think this will ever be true so we can probably skip this check.

Reply:
Probably. When I was testing I had an empty object as the `snapshot` to the player. It made the function return `NaN`. But if's unnecessary in reality I can remove it.